data-resource/src/blake3.rs

```diff
@@ -0,0 +1,84 @@
+use crate::ResourceIdTrait;
+use data_error::Result;
+use serde::{Deserialize, Serialize};
+use std::{
+    fs,
+    io::{BufRead, BufReader},
+    path::Path,
+};
+
+use blake3::Hasher;
+use hex::encode;
+
+/// Represents a resource identifier using the BLAKE3 algorithm.
+///
+/// Uses [`blake3`] crate to compute the hash value.
+#[derive(
+    Eq,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    Hash,
+    Clone,
+    Copy,
+    Debug,
+    Deserialize,
+    Serialize,
+)]
+pub struct ResourceId {}
+
+impl ResourceIdTrait for ResourceId {
+    type HashType = String;
```

Comment:
Probably `Vec<u64>` would be better.

Reply:
We can't directly use `Vec<u8>` since it doesn't come with `Display` or `FromStr` implementations, which we typically expect `ResourceId` to have in other crates. To handle this, we can wrap it in a struct
```rust
#[derive(
    Hash, Ord, PartialOrd, Eq, PartialEq, Clone, Debug, Serialize, Deserialize,
)]
pub struct Hash(Vec<u8>);

impl Display for Hash {
}

impl FromStr for Hash {
}
```
I gave it a try in this commit https://github.com/ARK-Builders/ark-rust/commit/382aa9bb51e8c0b4622a5a08988a31c10496796d

---
Ran benchmarks comparing it to using a plain `String` for the hash (hex encoded), and here are the results:
```sh
Gnuplot not found, using plotters backend
resource_id_creation/compute_from_path:../test-assets/lena.jpg
                        time:   [194.53 µs 195.35 µs 196.66 µs]
                        change: [+0.0968% +0.4659% +1.0237%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  3 (3.00%) high severe
resource_id_creation/compute_from_path:../test-assets/test.pdf
                        time:   [1.6235 ms 1.6256 ms 1.6287 ms]
                        change: [+0.2701% +0.4232% +0.5821%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe
resource_id_creation/compute_from_bytes:small
                        time:   [1.2565 µs 1.2570 µs 1.2576 µs]
                        change: [-9.3952% -9.2482% -9.1129%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) low severe
  5 (5.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe
resource_id_creation/compute_from_bytes:medium
                        time:   [37.043 µs 37.055 µs 37.067 µs]
                        change: [-0.7187% -0.4734% -0.2778%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  6 (6.00%) high mild
  2 (2.00%) high severe
resource_id_creation/compute_from_bytes:large
                        time:   [592.64 µs 592.84 µs 593.05 µs]
                        change: [-2.1562% -1.2193% -0.5402%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  3 (3.00%) high mild
  1 (1.00%) high severe
```
The performance seems relatively unchanged.

  
If you still prefer storing the hash in `Vec<u8>` for other reasons, I can directly cherrypick the commit to this PR.

Reply:
@tareknaser benchmarking is great. If it doesn't reveal any performance benefit then String is fine.

Could you also benchmark with `Vec<u64>` instead of `Vec<u8>`? Only `compute_from_bytes ` benchmarks is enough.

Reply:
This commit modifies the data type from `Vec<u8>` to `Vec<u64>`: https://github.com/tareknaser/ark-rust/commit/8f0b3473ae567b303e4b8b21edfba199054e8c28 
Here are the benchmarks for moving from hash of type `String` to `Vec<u64>`
```sh
resource_id_creation/compute_from_bytes:small
                        time:   [241.08 ns 241.85 ns 242.89 ns]
                        change: [+0.0702% +0.5050% +0.9694%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild
  4 (4.00%) high severe
resource_id_creation/compute_from_bytes:medium
                        time:   [15.043 µs 15.088 µs 15.148 µs]
                        change: [-0.2250% +0.3452% +1.1816%] (p = 0.35 > 0.05)
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  1 (1.00%) high mild
  10 (10.00%) high severe
resource_id_creation/compute_from_bytes:large
                        time:   [246.52 µs 247.14 µs 248.10 µs]
                        change: [-0.4475% +0.0280% +0.4609%] (p = 0.91 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  3 (3.00%) high mild
  6 (6.00%) high severe

     Running benches/index_build_benchmark.rs (target/release/deps/index_build_benchmark-6ce1452d366f8dbc)
Gnuplot not found, using plotters backend
index_build/index_build/../test-assets/
                        time:   [1.0621 ms 1.0648 ms 1.0686 ms]
                        change: [-0.3661% -0.0233% +0.3024%] (p = 0.89 > 0.05)
                        No change in performance detected.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low severe
  2 (2.00%) high mild
  6 (6.00%) high severe
```
There is no change in performance. 

Do you think we should stick with `String` since it's simpler to read and maintain?

Reply:
@tareknaser could you perform last benchmarks with 10mb data size? But overall it seems that `Vec<u64>` is a premature optimization. Maybe it's different on really huge files like 1GB but we target on 512kb - 15mb range as the most common.

Reply:
Sure. 
I ran some benchmarks on a 13 MB PDF file, using commit 8f0b347.
Here are the benchmarks for the transition from a hash of type `String` to `Vec<u64>`.

```sh
tareknasser@Tareks-MacBook ark-rust % ls -l /Users/tareknasser/Documents/books/Database_System_Concepts.pdf
-rw-r--r--@ 1 tareknasser  staff  13802981 Dec 28 20:15 /Users/tareknasser/Documents/books/Database_System_Concepts.pdf
tareknasser@Tareks-MacBook ark-rust % cargo bench 
   Compiling data-resource v0.1.0 (/Users/tareknasser/Documents/workspace/rust/arkbuilders/ark-rust/data-resource)
   Compiling fs-index v0.1.0 (/Users/tareknasser/Documents/workspace/rust/arkbuilders/ark-rust/fs-index)
    Finished bench [optimized] target(s) in 2.41s
     Running benches/id_creation_benchmarks.rs (target/release/deps/bench_resource_id_creation-69c86f48219418ac)
Gnuplot not found, using plotters backend
Benchmarking resource_id_creation/compute_from_path:../test-assets/lena.jpg: Collecting 100 samples in estimated 20.147 s (232kresource_id_creation/compute_from_path:../test-assets/lena.jpg
                        time:   [86.469 µs 86.533 µs 86.612 µs]
                        change: [-0.9975% -0.6224% -0.3123%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe
Benchmarking resource_id_creation/compute_from_path:/Users/tareknasser/Documents/books/Database_System_Concepts.p...: Warming uBenchmarking resource_id_creation/compute_from_path:/Users/tareknasser/Documents/books/Database_System_Concepts.p...: Collectinresource_id_creation/compute_from_path:/Users/tareknasser/Documents/books/Database_System_Concepts.p...
                        time:   [9.5751 ms 9.5894 ms 9.6052 ms]
                        change: [-1.4499% -0.9298% -0.4874%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe
```
Performance seems to be unchanged.

Reply:
Thanks! Let's stick to `String` then